### PR TITLE
Add capsule exec guide

### PR DESCRIPTION
**Pull request by mohanson**

### CODE REVIEW

**Comment by sopium** [comment]
We should probably release a new version of ckb-std and capsule and add an example using the new entrypoint and argv api.

File: website/docs/labs/capsule-exec.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
+---
+id: capsule-exec
+title: Exec syscall in Capsule
+---
+
+## Introduction
+
+[Exec](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0034-vm-syscalls-2/0034-vm-syscalls-2.md#exec) is a new syscall provided by [ckb2021](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0037-ckb2021/0037-ckb2021.md). To understand what exec syscall does, we recommend you read this article first: [https://man7.org/linux/man-pages/man3/exec.3.html](https://man7.org/linux/man-pages/man3/exec.3.html). In short: Exec runs an executable file from specified cell data in the context of an already existing vm, replacing the previous executable. The used cycles does not change, but the code, registers and memory of the vm are replaced by those of the new program. This means, **control flow will never return to the main script**.
```
Comment (linnnsss) [inline website/docs/labs/capsule-exec.md:8]: [Exec](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0034-vm-syscalls-2/0034-vm-syscalls-2.md#exec) is a new syscall provided by [ckb2021](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0037-ckb2021/0037-ckb2021.md). To understand what exec syscall does, we recommend reading [this article] (https://man7.org/linux/man-pages/man3/exec.3.html) first. In short: Exec runs an executable file from specified cell data in the context of an already existing VM, replacing the previous executable. The used cycles do not change, but the code, registers, and memory of the VM are replaced by those of the new program, meaning **control flow will never return to the main script**.

File: website/docs/labs/capsule-exec.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+You can imagine exec as a router. When some conditions are met, the main script will completely hand over control to a certain sub-script.
+
+```text
+              ┌--> if State1 --> Exec(Sub-script1)
+Main script --+--> if State2 --> Exec(Sub-script2)
+              └--> if State3 --> Exec(Sub-script3)
+```
+
+Compare with [Dynamic libraries](https://docs.nervos.org/docs/labs/capsule-dynamic-loading-tutorial/), exec has the following significant advantages:
```
Comment (linnnsss) [inline website/docs/labs/capsule-exec.md:18]: Compared with

File: website/docs/labs/capsule-exec.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+New contract "always-failure"
+     Created binary (application) `always-failure` package
+Rewrite Cargo.toml
+Rewrite capsule.toml
+Done
+```
+
+</details>
+
+## Write Always failure sub-script
```
Comment (linnnsss) [inline website/docs/labs/capsule-exec.md:80]: Write `always-failure` sub-script

File: website/docs/labs/capsule-exec.md [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+Put the following codes into `contracts/always-failure/main.rs`. As you can see, the script always returns 42, which means that if the script is used as a lock script, the cell will never be unlocked.
```
Comment (linnnsss) [inline website/docs/labs/capsule-exec.md:82]: the following code

File: website/docs/labs/capsule-exec.md [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+```rs
+fn program_entry(_argc: u64, _argv: *const *const u8) -> i8 {
+    return 42;
+}
+```
+
+## Write exec demo script
+
+Put the following codes into `contracts/ckb-exec-demo/main.rs`.
```
Comment (linnnsss) [inline website/docs/labs/capsule-exec.md:92]: the following code

File: website/docs/labs/capsule-exec.md [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    let r = exec(0, Source::CellDep, 0, 0, &[]);
+    if r != 0 {
+        // Call exec syscall failed.
+        return 10;
+    }
+    return 0;
+}
+```
+
+This script does only one thing, when executing `exec(0, Source::CellDep, 0, 0, &[])`, CKB-VM will look for the first dep_cell, and execute the code in it.
```
Comment (linnnsss) [inline website/docs/labs/capsule-exec.md:105]: ... does only one thing: When

File: website/docs/labs/capsule-exec.md [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+              ┌--> if State1 --> Exec(Sub-script1)
+Main script --+--> if State2 --> Exec(Sub-script2)
+              └--> if State3 --> Exec(Sub-script3)
+```
+
+Compared with [Dynamic libraries](https://docs.nervos.org/docs/labs/capsule-dynamic-loading-tutorial/), exec has the following significant advantages:
+
+- All sub-scripts are complete scripts. They can be used alone, or they can be called by exec.
+- Sub-scripts have a separate 4M memory space.
+
```
Comment (XuJiandong) [inline website/docs/labs/capsule-exec.md:22]: also add disadvantages: 1. never return 2. hard to exchanged data between scripts

**Review by XuJiandong — CHANGES_REQUESTED** [review]

**Comment by phroi** [comment]
Just noticed this super interesting docs pull request, this phrase in particular caught my eye:
> All sub-scripts are complete scripts. They can be used alone, or they can be called by exec.

For example does this entail I can directly call Exec Omnilock from my scripts?
If so, how should the arguments be formatted? Could you make a Rust example on how to call Omnilock with Exec?

**Comment by mohanson** [comment]
> Just noticed this super interesting docs pull request, this phrase in particular caught my eye:
> 
> > All sub-scripts are complete scripts. They can be used alone, or they can be called by exec.
> 
> For example does this entail I can directly call Exec Omnilock from my scripts? If so, how should the arguments be formatted? Could you make a Rust example on how to call Omnilock with Exec?

I updated this draft to show how to pass in and use parameters.

Please read the latest draft and demo: https://github.com/mohanson/ckb-exec-demo

**Comment by phroi** [comment]
Much better, thanks!! I still had some doubts about the `place` argument, but after looking up [the `Exec` implementation](https://github.com/nervosnetwork/ckb/blob/develop/script/src/syscalls/exec.rs) I got it.

So to reply to my own question: I don't believe it's possible to use `OmniLock` with `Exec` unless the implementation is changed to account for this new use.

**Pull request merged**